### PR TITLE
ci: Add pre-commit action

### PR DESCRIPTION
**Pull request by Techassi**
This adds a re-usable pre-commit action. It optionally installs the Rust toolchain and Hadolint in the requested version.

### CODE REVIEW

File: setup-pre-commit/action.yml [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+      uses: actions/setup-python@f677139bbe7f9c59b41e40162b753c062f5d49a3 # v5.2.0
+      with:
+        python-version: ${{ inputs.python-version }}
+
+    - name: Setup Rust Toolchain
+      uses: dtolnay/rust-toolchain@7b1c307e0dcbda6122208f10795a713336a9b35a
+      if: ${{ inputs.rust }}
+      with:
+        toolchain: ${{ inputs.rust }}
+        components: rustfmt,clippy
```
Comment (NickLarsenNZ) [inline setup-pre-commit/action.yml:28]: Nothing to change now, but maybe this should be overridable. For example, somewhere, we needed `rust-src` too. Maybe we want to add more components one day?

**Review by NickLarsenNZ — COMMENTED** [review]
A comment about the cargo action

**Comment by NickLarsenNZ** [comment]
Also, the readme will need to have a note about needing to adjust the fetch-depth on checkout.

File: setup-pre-commit/action.yml (hunk 1 above)
Comment (Techassi) [reply setup-pre-commit/action.yml]: Added in 69b8e50

File: .yamllint.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+---
+extends: default
+
+ignore: |
+  deploy/helm/**/templates
+
```
Comment (NickLarsenNZ) [inline .yamllint.yaml:6]: I think this is leftover from copy/paste

```suggestion
```

File: run-pre-commit/action.yml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+
+        mkdir -p "$LOCATION_DIR"
+        curl -sL -o "${LOCATION_BIN}" "https://github.com/hadolint/hadolint/releases/download/${{ inputs.hadolint }}/hadolint-$SYSTEM-$ARCH"
+        chmod 700 "${LOCATION_BIN}"
+
+        echo "$LOCATION_DIR" >> "$GITHUB_PATH"
+
+    - uses: pre-commit/action@2c7b3805fd2a0fd8c1884dcaebf91fc102a13ecd # v3.0.1
+      with:
+        extra_args: "--from-ref ${{ github.event.pull_request.base.sha }} --to-ref ${{ github.event.pull_request.head.sha }}"
```
Comment (NickLarsenNZ) [inline run-pre-commit/action.yml:55]: This is probably ok, as the values should always exist and have no spaces, but for extra safety, we could quote the variable parameters:

```suggestion
        extra_args: "--from-ref '${{ github.event.pull_request.base.sha }}' --to-ref '${{ github.event.pull_request.head.sha }}'"
```

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+    hooks:
+      - id: yamllint
+
+  - repo: https://github.com/igorshubovych/markdownlint-cli
+    rev: f295829140d25717bc79368d3f966fc1f67a824f # 0.41.0
+    hooks:
+      - id: markdownlint
+        types: [text]
+        files: \.md(\.j2)*$
+
```
Comment (NickLarsenNZ) [inline .pre-commit-config.yaml:26]: We don't have j2 files here, so we can drop the extra config:

```suggestion
```

File: run-pre-commit/action.yml [hunk 5]
```diff
@@ -0,0 +1,55 @@
+---
+name: Run pre-commit
+description: |
+  This actions sets up Python, the pre-commit tool and additional tools required
+  for various hooks. It then runs pre-commit against the changed files. This
+  actions expects checkouts with depth 0.
```
Comment (NickLarsenNZ) [inline run-pre-commit/action.yml:6]:
```suggestion
  This actions sets up Python, the pre-commit tool, and additional tools 
  required for various hooks. It then runs pre-commit against the changed files.
  This actions expects checkouts with depth 0.
```

**Review by NickLarsenNZ — CHANGES_REQUESTED** [review]
A couple of changes, the rest are optional.

File: .yamllint.yaml (hunk 2 above)
Comment (Techassi) [reply .yamllint.yaml:6]: Yupp, indeed.

File: run-pre-commit/action.yml (hunk 3 above)
Comment (Techassi) [reply run-pre-commit/action.yml:55]: Yes, makes sense.

**Review by NickLarsenNZ — DISMISSED** [review]
LGTM

**Review by NickLarsenNZ — APPROVED** [review]
LGTM

**Pull request merged**